src/controllers/orderController.js

```diff
@@ -5,58 +5,103 @@ const { BadRequestError } = require('../errors');
-  }
 
-  // Create an Order document in the database
-  const order = await Order.create({
-    user: req.user.userId,
-    orderItems,
-    subtotal,
-    tax,
-    total,
-  });
 
-  // Create a payment intent with Stripe
-  const paymentIntent = await stripe.paymentIntents.create({
-    amount: total * 100, // Stripe requires amount in cents
-    currency: 'usd',
-    description: `Order #${order._id}`,
-    metadata: {
-      userId: req.user.userId,
-      orderId: order._id, // key-values
-      totalQuantity: orderItems.reduce(
-        (total, item) => total + item.quantity,
-        0
-      ), // the total number of items in the order
-      totalAlbums: orderItems.length, // the total number of albums purchased in order
-    },
-  });
 
-  // Save the paymentIntent to the Order model
-  order.paymentIntentId = paymentIntent.id;
-  await order.save();
 
-  // Create PurchasedAlbum entries for each album in orderItems
-  for (const orderItem of orderItems) {
-    await PurchasedAlbum.create({
-      album: orderItem.album, // album ID from orderItem
-      user: req.user.userId, // user ID from the request
+// Populate the orderItems with full album data
+const populatedOrderItems = await Promise.all(
+  orderItems.map(async (orderItem) => {
+    return await orderItem.populate('album').execPopulate();
+  })
+);
+  // Create an array to store the order items with full album data
+  const orderItemsWithFullAlbum = populatedOrderItems.map((orderItem) => ({
+    album: orderItem.album.toObject(), // Use toObject() to convert the full album to a plain JavaScript object
+    quantity: orderItem.quantity,
+  }));
+
+
+
+
+    const order = new Order({
+      user: req.user.userId,
+      orderItems: orderItemsWithFullAlbum,// Use the array with full album data
+      subtotal,
+      tax,
+      total,
     });
-  }
 
-  // Send the payment intent client secret and order information to the client
-  res
-    .status(StatusCodes.CREATED)
-    .json({ clientSecret: paymentIntent.client_secret, order });
+     await order.save({ session });
+
+   
+    const paymentIntent = await stripe.paymentIntents.create({
+      amount: total * 100,
+      currency: 'usd',
+      description: `Order #${order._id}`,
+      metadata: {
+        userId: req.user.userId,
+        orderId: order._id,
+        totalQuantity: orderItems.reduce(
+          (total, item) => total + item.quantity,
+          0
+        ),
+        totalAlbums: orderItems.length,
+      },
+    });
+
+    order.paymentIntentId = paymentIntent.id;
+
+    for (const orderItem of orderItems) {
+      const purchasedAlbum = new PurchasedAlbum({
+        album: orderItem.album,
+        user: req.user.userId,
+      });
+
+      await purchasedAlbum.save({ session });
+    }
+
+
+
+
+    await session.commitTransaction();
+
+    
+    // Send the order completion email after THE ORDER HAS STATUS COMPLETE
+    await sendOrderCompletedEmail(user.email, user.name, orderItemsWithFullAlbum);
```

Comment:
If this is where we need the full album info in the order item, we can populate it from the saved order:

`const orderItemsWithFullAlbum = await order.populate({ path: 'orderItems.album'})`

Reply:
And the email function is currently expecting four arguments (to, username, orderitems, and total), so we'll want to pass in something like:

```
await sendOrderCompletedEmail(user.email, user.username, orderItemsWithFullAlbum.orderItems, orderItemsWithFullAlbum.total);
```

Reply:
@akosasante not sure what I'm doing wrong but trying this way I still  don't get album populated 
![image](https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/103755090/561af3fe-dc86-4d3a-ab11-6e448c431c28)
![image](https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/103755090/ac87187a-50f6-4e2f-8535-e590f26bc965)

Reply:
also noticed that if i create several orders with complete status- it sends email only to the oldest one  (I was testing it by sending different quantites)